### DISCUSSION TITLE
Why does this compile? (Possible unsoundness?)

### DISCUSSION THREAD

**Discussion by SUPERCILEX** [tokio-rs/tokio#4597] [Q&A]
**Version**

`tokio v1.17.0`
`rustc 1.61.0-nightly (76d770ac2 2022-04-02)`

**Description**

```rust
async fn test() {
    let s = UdpSocket::bind("0.0.0.0:12235").await.unwrap();
    let mut rng = thread_rng();

    loop {
        let mut bytes = [0; 10];
        let (len, addr) = s.recv_from(&mut bytes).await.unwrap();
        s.send_to(&bytes[..rng.gen::<usize>() % len], addr)
            .await
            .unwrap();
    }
}
```

I'm probably missing something, but I feel like this shouldn't compile. `ThreadRng` is not `Send` so shouldn't it fail to cross the `await` point? I thought you could end up on any thread after hitting an `await` which would break the `ThreadRng`.

**Comment by Darksonn** [comment]
It's only an issue once you actually spawn it on a runtime on which it can be moved across threads. So, the error will happen at the `tokio::spawn` call.

If you only use it from your `#[tokio::main]`, then it will compile because the main function cannot be moved across threads.

**Reply by SUPERCILEX** [reply]
Ohhh, interesting. I found [this code](https://github.com/tokio-rs/tokio/blob/f8a6cf49cd5a7a9436790c486b2879def138adc3/tokio/src/park/thread.rs#L267). As I understand it, await calls cause the main thread to park/block/sleep/whatever until the reactor presumably unparks it with the waker once things are ready. Is my understanding correct?


### Appendix: repro

```rust
fn root1() {
    let runtime = tokio::runtime::Builder::new_multi_thread().build().unwrap();
    runtime.block_on(test()); // WORKS
}

fn root2() {
    let runtime = tokio::runtime::Builder::new_multi_thread().build().unwrap();
    runtime.block_on(async {
        tokio::spawn(test()); // FAILS
    });
}

async fn test() {
    let s = UdpSocket::bind("0.0.0.0:12235").await.unwrap();
    let mut rng = thread_rng();

    loop {
        let mut bytes = [0; 10];
        let (len, addr) = s.recv_from(&mut bytes).await.unwrap();
        s.send_to(&bytes[..rng.gen::<usize>() % len], addr)
            .await
            .unwrap();
    }
}
```
